src/ChunkedStreamDecoder.php

```diff
@@ -0,0 +1,174 @@
+    protected $maxBufferLength;
+
+    /**
+     * @var bool
+     */
+    protected $nextChunkIsLength = true;
+
+    /**
+     * @var ReadableStreamInterface
+     */
+    protected $stream;
+
+    /**
+     * @param ReadableStreamInterface $stream
+     * @param int $maxBufferLength
+     */
+    public function __construct(ReadableStreamInterface $stream, $maxBufferLength = self::MAX_BUFFER_LENGTH)
+    {
+        $this->stream = $stream;
+        $this->stream->on('data', array($this, 'handleData'));
+        Util::forwardEvents($this->stream, $this, [
+            'error',
+        ]);
+        $this->maxBufferLength = $maxBufferLength;
+    }
+
+    /** @internal */
+    public function handleData($data)
+    {
+        $this->buffer .= $data;
+
+        do {
+            $bufferLength = strlen($this->buffer);
+            $continue = $this->iterateBuffer();
+            $iteratedBufferLength = strlen($this->buffer);
+        } while (
+            $continue &&
+            $bufferLength !== $iteratedBufferLength &&
+            $iteratedBufferLength > 0 &&
+            strpos($this->buffer, static::CRLF) !== false &&
+            $iteratedBufferLength <= $this->maxBufferLength
+        );
+
+        if ($iteratedBufferLength >= $this->maxBufferLength) {
+            $this->emit('error', [
+                new Exception('The current buffer is longer then the ' . $this->maxBufferLength / 1024 . 'KB we have set as maximum'),
+            ]);
+            $this->close();
+            return false;
+        }
+    }
+
+    protected function iterateBuffer()
+    {
+        if (strlen($this->buffer) <= 1) {
+            return false;
+        }
+
+        if ($this->nextChunkIsLength) {
+            $crlfPosition = strpos($this->buffer, static::CRLF);
+            if ($crlfPosition === false) {
+                return false; // Chunk header hasn't completely come in yet
+            }
+            $this->nextChunkIsLength = false;
+            $lengthChunk = substr($this->buffer, 0, $crlfPosition);
+            if (strpos($lengthChunk, ';') !== false) {
+                list($lengthChunk) = explode(';', $lengthChunk, 2);
+            }
+            if (dechex(hexdec($lengthChunk)) !== $lengthChunk) {
+                $this->emit('error', [
+                    new Exception('Unable to validate "' . $lengthChunk . '" as chunk length header"'),
+                ]);
+                $this->close();
+                return false;
+            }
+            $this->remainingLength = hexdec($lengthChunk);
+            if ($this->remainingLength >= $this->maxBufferLength) {
+                $this->emit('error', [
+                    new Exception('Expected chunk length is longer then the ' . $this->maxBufferLength / 1024 . 'KB we have set as maximum'),
+                ]);
+                $this->close();
+                return false;
+            }
+            $this->buffer = substr($this->buffer, $crlfPosition + 2);
+            return true;
```

Comment:
If the underlying stream emits an `end` while we're expecting more data, this actually means the chunked stream should emit an `error`, because it contains a chunk that is now invalid.

If the chunk decoder sees a zero length chunk it means the chunked stream ends, it should no longer expect any data from the input stream. (This is in fact also used for connection reuse, as the underlying streams can be reused for the next request/response)

See for example https://github.com/clue/php-term-react/blob/bb8784283aef16b8b1fe3ed5a8fe16d4669cfc92/src/ControlCodeParser.php#L205